src/consensus.rs

```diff
@@ -822,92 +760,228 @@ fn handle_committed_entries(
-    // Use dedicated transport channel for who_is because of specific timeout
-    let channel = make_grpc_channel(
-        bootstrap_timeout,
-        bootstrap_timeout,
-        bootstrap_uri,
-        tls_config,
-    )
-    .await
-    .map_err(|err| anyhow!("Failed to create timeout channel: {}", err))?;
-    let mut client = RaftClient::new(channel);
-    Ok(client
-        .who_is(tonic::Request::new(GrpcPeerId { id: peer_id }))
-        .await?
-        .into_inner()
-        .uri
-        .parse()?)
+    consensus_config: Arc<ConsensusConfig>,
+    consensus_state: ConsensusStateRef,
+    transport_channel_pool: Arc<TransportChannelPool>,
+}
+
+impl RaftMessageBroker {
+    pub fn new(
+        runtime: Handle,
+        bootstrap_uri: Option<Uri>,
+        tls_config: Option<ClientTlsConfig>,
+        consensus_config: ConsensusConfig,
+        consensus_state: ConsensusStateRef,
+        transport_channel_pool: Arc<TransportChannelPool>,
+    ) -> Self {
+        Self {
+            senders: HashMap::new(),
+            runtime,
+            bootstrap_uri,
+            tls_config,
+            consensus_config: consensus_config.into(),
+            consensus_state,
+            transport_channel_pool,
+        }
+    }
+
+    pub fn send(&mut self, messages: impl IntoIterator<Item = RaftMessage>) {
+        for message in messages {
+            let peer_id = message.to;
+
+            let sender = match self.senders.get(&peer_id) {
+                Some(sender) => sender,
+                None => {
+                    log::debug!("Spawning message sender task for peer {peer_id}...");
+
+                    let (task, handle) = self.message_sender();
+                    let future = self.runtime.spawn(task.exec());
+                    drop(future); // drop `JoinFuture` explicitly to make clippy happy
+
+                    self.senders.insert(peer_id, handle);
+
+                    self.senders
+                        .get(&peer_id)
+                        .expect("message sender task spawned")
+                }
+            };
+
+            match sender.try_send(message) {
+                Ok(()) => (),
+
+                Err(tokio::sync::mpsc::error::TrySendError::Full(message)) => {
+                    if log::max_level() >= log::Level::Debug {
+                        log::error!(
+                            "Failed to forward message {message:?} to message sender task {peer_id}: \
+                             message sender taks queue is full"
+                        );
+                    } else {
+                        log::error!(
+                            "Failed to forward message to message sender task {peer_id}: \
+                             message sender task queue is full"
+                        );
+                    }
+                }
+
+                Err(tokio::sync::mpsc::error::TrySendError::Closed(message)) => {
+                    panic!(
```

Comment:
If we panic here we are taking down the whole consensus thread.
Why panic over logging?

Reply:
We could even try to recover by restarting the `RaftMessageSender`. And I even drafted the code to do it... but the code with recovery handling is less readable, than the "plain" one. :/

Reply:
Replaced `panic` with `unreachable`.